### ISSUE TITLE
Trouble with tutorial example 

### ISSUE THREAD

**Issue by gtatters** [rstudio/keras#985]
Hi there,
I am trying to get the tutorial example from the blog to work (https://blog.rstudio.com/2017/09/05/keras-for-r/) but have hit a snag.  

Using the following code, the datasets load fine, but I cannot issue the dim function in the final line, as a get a dimension error

```
library(keras)
mnist <- dataset_mnist()
x_train <- mnist$train$x
y_train <- mnist$train$y
x_test <- mnist$test$x
y_test <- mnist$test$y
dim(x_train) <- c(nrow(x_train), 784)
```

Error code:
Error in dim(x_train) <- c(nrow(x_train), 784) : invalid first argument

The 4 variables exist as numpy.ndarray objects

Specifications:
Running on Mac OSX (10.14.6), Rstudio / R (3.6.2)

I suspect it is something I've not installed properly, but there I thought I followed the instructions in the tutorial.  

Are you able to offer any advice?

Thanks

**Comment by skeydan** [comment]
What does 

```
tensorflow::tf_config()
```

say? And what is your `sessionInfo()`?

**Comment by gtatters** [comment]
For the first I get:

TensorFlow v2.0.0 (~/.virtualenvs/r-reticulate/lib/python3.7/site-packages/tensorflow)
Python v3.7 (~/.virtualenvs/r-reticulate/bin/python)

For the sessionInfo():

R version 3.6.1 (2019-07-05)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Mojave 10.14.6

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_CA.UTF-8/en_CA.UTF-8/en_CA.UTF-8/C/en_CA.UTF-8/en_CA.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] tensorflow_2.0.0 keras_2.2.5.0   

loaded via a namespace (and not attached):
 [1] compiler_3.6.1  magrittr_1.5    R6_2.4.1        generics_0.0.2 
 [5] tools_3.6.1     whisker_0.4     base64enc_0.1-3 Rcpp_1.0.3     
 [9] reticulate_1.14 zeallot_0.1.0   jsonlite_1.6.1  tfruns_1.4

**Comment by gtatters** [comment]
I'm not sure why the line is crossed out above.  I did not add that.

**Comment by skeydan** [comment]
I can only guess, - perhaps an issue with the Python on your system? (I remember seeing problems in some Mac versions.) Would you mind installing Miniconda

https://docs.conda.io/en/latest/miniconda.html

and trying to install `tensorflow` there?

```
install_tensorflow(method = "conda")
```
In general, using Miniconda is now the "most" recommended way when binding to Python via `reticulate`, certainly on Windows and probably on Mac as well.

**Comment by gtatters** [comment]
ok I will try that.  the Keras website didn't specify any other external install requirements.  Is there a version requirement for python?  I initially had miniconda installed (both 2.7 and 3.7) and the tutorial still failed at this line.  I can try again.  Will post once I've tried.

**Comment by gtatters** [comment]
so when I type insall_tensorflow(method="conda"), after having checked that I have miniconda2 installed, I get a long error message.  I really do not understand these error messages. But does this sound like a reticulate problem? 

Using virtual environment '/Users/MYNAME/.virtualenvs/r-reticulate' ...
Requirement already up-to-date: tensorflow==2.0.0 in /Users/MYNAME/.virtualenvs/r-reticulate/lib/python3.7/site-packages (2.0.0)
Requirement already satisfied, skipping upgrade: termcolor>=1.1.0 in /Users/MYNAME/.virtualenvs/r-reticulate/lib/python3.7/site-packages (from tensorflow==2.0.0) (1.1.0)

**Comment by gtatters** [comment]
Not having much success.  Does anyone have a clear set of instructions to get all the packages installed?

**Comment by skeydan** [comment]
What is that error message you mention?

Did you restart the R session before the above 
```
install_tensorflow(method = "conda")
```
?
If not, can you try again?

Normally that virtualenv should not be used when you issue that command...
You could also try creating a conda environment upfront, activating it, and only then start RStudio and install TensorFlow.

**Comment by gtatters** [comment]
ok. so to track what I'm doing (since the instructions assume user knowledge I lack). 
I have removed all versions of miniconda and reinstalled miniconda 3 manually too my user folder (~/opt/miniconda3)

then simply run the following in RStudio:

```
library(keras)
library(tensorflow)
install_tensorflow(method = "conda")
```

And the rest is the output I get from RStudio....


Collecting package metadata (current_repodata.json): ...working... done
Solving environment: ...working... done

## Package Plan ##

  environment location: /Users/MYUSERNAME/opt/miniconda3/envs/r-reticulate

  added / updated specs:
    - python=3.6


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2020.1.1   |                0         125 KB
    certifi-2019.11.28         |           py36_0         156 KB
    openssl-1.1.1d             |       h1de35cc_4         2.2 MB
    pip-20.0.2                 |           py36_1         1.7 MB
    python-3.6.10              |       h359304d_0        17.1 MB
    setuptools-45.2.0          |           py36_0         511 KB
    sqlite-3.31.1              |       ha441bb4_0         1.2 MB
    wheel-0.34.2               |           py36_0          50 KB
    ------------------------------------------------------------
                                           Total:        23.1 MB

The following NEW packages will be INSTALLED:

  ca-certificates    pkgs/main/osx-64::ca-certificates-2020.1.1-0
  certifi            pkgs/main/osx-64::certifi-2019.11.28-py36_0
  libcxx             pkgs/main/osx-64::libcxx-4.0.1-hcfea43d_1
  libcxxabi          pkgs/main/osx-64::libcxxabi-4.0.1-hcfea43d_1
  libedit            pkgs/main/osx-64::libedit-3.1.20181209-hb402a30_0
  libffi             pkgs/main/osx-64::libffi-3.2.1-h475c297_4
  ncurses            pkgs/main/osx-64::ncurses-6.1-h0a44026_1
  openssl            pkgs/main/osx-64::openssl-1.1.1d-h1de35cc_4
  pip                pkgs/main/osx-64::pip-20.0.2-py36_1
  python             pkgs/main/osx-64::python-3.6.10-h359304d_0
  readline           pkgs/main/osx-64::readline-7.0-h1de35cc_5
  setuptools         pkgs/main/osx-64::setuptools-45.2.0-py36_0
  sqlite             pkgs/main/osx-64::sqlite-3.31.1-ha441bb4_0
  tk                 pkgs/main/osx-64::tk-8.6.8-ha441bb4_0
  wheel              pkgs/main/osx-64::wheel-0.34.2-py36_0
  xz                 pkgs/main/osx-64::xz-5.2.4-h1de35cc_4
  zlib               pkgs/main/osx-64::zlib-1.2.11-h1de35cc_3



Downloading and Extracting Packages
openssl-1.1.1d       | 2.2 MB    | ########## | 100% 
python-3.6.10        | 17.1 MB   | ########## | 100% 
ca-certificates-2020 | 125 KB    | ########## | 100% 
sqlite-3.31.1        | 1.2 MB    | ########## | 100% 
wheel-0.34.2         | 50 KB     | ########## | 100% 
pip-20.0.2           | 1.7 MB    | ########## | 100% 
certifi-2019.11.28   | 156 KB    | ########## | 100% 
setuptools-45.2.0    | 511 KB    | ########## | 100% 
Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
#
# To activate this environment, use
#
#     $ conda activate r-reticulate
#
# To deactivate an active environment, use
#
#     $ conda deactivate



==> WARNING: A newer version of conda exists. <==
  current version: 4.7.12
  latest version: 4.8.2

Please update conda by running

    $ conda update -n base -c defaults conda


Error in sitecustomize; set PYTHONVERBOSE for traceback:
KeyError: 'PYTHONPATH'
Collecting tensorflow==2.0.0
  Using cached https://files.pythonhosted.org/packages/c8/a1/2ab46a175c916b0149ccb9edc06202bce6365455779fa251c1f59a4c7806/tensorflow-2.0.0-cp36-cp36m-macosx_10_11_x86_64.whl
Processing /Users/MYUSERNAME/Library/Caches/pip/wheels/2c/b1/94/43d03e130b929aae7ba3f8d15cbd7bc0d1cb5bb38a5c721833/opt_einsum-3.1.0-cp36-none-any.whl
Collecting grpcio>=1.8.6
  Using cached https://files.pythonhosted.org/packages/e2/80/e64d6a7648e44ea210aec5f2b8b0a0b08f15a6cfe5fe32361b80b42df8a2/grpcio-1.27.2-cp36-cp36m-macosx_10_9_x86_64.whl
Processing /Users/MYUSERNAME/Library/Caches/pip/wheels/8e/28/49/fad4e7f0b9a1227708cbbee4487ac8558a7334849cb81c813d/absl_py-0.9.0-cp36-none-any.whl
Processing /Users/MYUSERNAME/Library/Caches/pip/wheels/54/f9/95/099544e9f879f719b14cf567fabb5aa7984263df0f025f3eef/wrapt-1.12.0-cp36-cp36m-macosx_10_9_x86_64.whl
Collecting keras-applications>=1.0.8
  Using cached https://files.pythonhosted.org/packages/71/e3/19762fdfc62877ae9102edf6342d71b28fbfd9dea3d2f96a882ce099b03f/Keras_Applications-1.0.8-py3-none-any.whl
Collecting keras-preprocessing>=1.0.5
  Using cached https://files.pythonhosted.org/packages/28/6a/8c1f62c37212d9fc441a7e26736df51ce6f0e38455816445471f10da4f0a/Keras_Preprocessing-1.1.0-py2.py3-none-any.whl
Collecting google-pasta>=0.1.6
  Using cached https://files.pythonhosted.org/packages/c3/fd/1e86bc4837cc9a3a5faf3db9b1854aa04ad35b5f381f9648fbe81a6f94e4/google_pasta-0.1.8-py3-none-any.whl
Collecting tensorboard<2.1.0,>=2.0.0
  Using cached https://files.pythonhosted.org/packages/76/54/99b9d5d52d5cb732f099baaaf7740403e83fe6b0cedde940fabd2b13d75a/tensorboard-2.0.2-py3-none-any.whl
Requirement already satisfied, skipping upgrade: wheel>=0.26 in /usr/local/lib/python2.7/site-packages (from tensorflow==2.0.0) (0.33.6)
Processing /Users/MYUSERNAME/Library/Caches/pip/wheels/5c/2e/7e/a1d4d4fcebe6c381f378ce7743a3ced3699feb89bcfbdadadd/gast-0.2.2-cp36-none-any.whl
Collecting protobuf>=3.6.1
  Using cached https://files.pythonhosted.org/packages/4c/65/3ac73d6a9f31de4b45ebff6885e9a4ccd16eccb63764dd406140d337fabd/protobuf-3.11.3-cp36-cp36m-macosx_10_9_x86_64.whl
Processing /Users/MYUSERNAME/Library/Caches/pip/wheels/7c/06/54/bc84598ba1daf8f970247f550b175aaaee85f68b4b0c5ab2c6/termcolor-1.1.0-cp36-none-any.whl
Collecting numpy<2.0,>=1.16.0
  Using cached https://files.pythonhosted.org/packages/82/63/eee643cc97f2bd22da87420f28fb6cd4b940c25f6eff6c4d2ca2e24a7022/numpy-1.18.1-cp36-cp36m-macosx_10_9_x86_64.whl
Collecting astor>=0.6.0
  Using cached https://files.pythonhosted.org/packages/c3/88/97eef84f48fa04fbd6750e62dcceafba6c63c81b7ac1420856c8dcc0a3f9/astor-0.8.1-py2.py3-none-any.whl
Collecting tensorflow-estimator<2.1.0,>=2.0.0
  Using cached https://files.pythonhosted.org/packages/fc/08/8b927337b7019c374719145d1dceba21a8bb909b93b1ad6f8fb7d22c1ca1/tensorflow_estimator-2.0.1-py2.py3-none-any.whl
Requirement already satisfied, skipping upgrade: six>=1.10.0 in /usr/local/Cellar/matplotlib/2.0.2/libexec/lib/python2.7/site-packages (from tensorflow==2.0.0) (1.10.0)
Collecting h5py
  Using cached https://files.pythonhosted.org/packages/ec/7f/a833846f5628d00f82ad87010f98294257535f1052e4a466888deba29f94/h5py-2.10.0-cp36-cp36m-macosx_10_6_intel.whl
Collecting google-auth<2,>=1.6.3
  Using cached https://files.pythonhosted.org/packages/5a/8d/e2ebbd0502627ed0d8a408162020e1c0792f088b49fddeedaaeebc206ed7/google_auth-1.11.2-py2.py3-none-any.whl
Collecting markdown>=2.6.8
  Using cached https://files.pythonhosted.org/packages/ab/c4/ba46d44855e6eb1770a12edace5a165a0c6de13349f592b9036257f3c3d3/Markdown-3.2.1-py2.py3-none-any.whl
Collecting werkzeug>=0.11.15
  Using cached https://files.pythonhosted.org/packages/ba/a5/d6f8a6e71f15364d35678a4ec8a0186f980b3bd2545f40ad51dd26a87fb1/Werkzeug-1.0.0-py2.py3-none-any.whl
Requirement already satisfied, skipping upgrade: setuptools>=41.0.0 in /usr/local/lib/python2.7/site-packages (from tensorboard<2.1.0,>=2.0.0->tensorflow==2.0.0) (41.4.0)
Collecting requests<3,>=2.21.0
  Using cached https://files.pythonhosted.org/packages/51/bd/23c926cd341ea6b7dd0b2a00aba99ae0f828be89d72b2190f27c11d4b7fb/requests-2.22.0-py2.py3-none-any.whl
Collecting google-auth-oauthlib<0.5,>=0.4.1
  Using cached https://files.pythonhosted.org/packages/7b/b8/88def36e74bee9fce511c9519571f4e485e890093ab7442284f4ffaef60b/google_auth_oauthlib-0.4.1-py2.py3-none-any.whl
Collecting pyasn1-modules>=0.2.1
  Using cached https://files.pythonhosted.org/packages/95/de/214830a981892a3e286c3794f41ae67a4495df1108c3da8a9f62159b9a9d/pyasn1_modules-0.2.8-py2.py3-none-any.whl
Collecting rsa<4.1,>=3.1.4
  Using cached https://files.pythonhosted.org/packages/02/e5/38518af393f7c214357079ce67a317307936896e961e35450b70fad2a9cf/rsa-4.0-py2.py3-none-any.whl
Collecting cachetools<5.0,>=2.0.0
  Using cached https://files.pythonhosted.org/packages/08/6a/abf83cb951617793fd49c98cb9456860f5df66ff89883c8660aa0672d425/cachetools-4.0.0-py3-none-any.whl
Requirement already satisfied, skipping upgrade: certifi>=2017.4.17 in /Users/MYUSERNAME/opt/miniconda3/envs/r-reticulate/lib/python3.6/site-packages (from requests<3,>=2.21.0->tensorboard<2.1.0,>=2.0.0->tensorflow==2.0.0) (2019.11.28)
Collecting urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1
  Using cached https://files.pythonhosted.org/packages/e8/74/6e4f91745020f967d09332bb2b8b9b10090957334692eb88ea4afe91b77f/urllib3-1.25.8-py2.py3-none-any.whl
Collecting idna<2.9,>=2.5
  Using cached https://files.pythonhosted.org/packages/14/2c/cd551d81dbe15200be1cf41cd03869a46fe7226e7450af7a6545bfc474c9/idna-2.8-py2.py3-none-any.whl
Collecting chardet<3.1.0,>=3.0.2
  Using cached https://files.pythonhosted.org/packages/bc/a9/01ffebfb562e4274b6487b4bb1ddec7ca55ec7510b22e4c51f14098443b8/chardet-3.0.4-py2.py3-none-any.whl
Collecting requests-oauthlib>=0.7.0
  Using cached https://files.pythonhosted.org/packages/a3/12/b92740d845ab62ea4edf04d2f4164d82532b5a0b03836d4d4e71c6f3d379/requests_oauthlib-1.3.0-py2.py3-none-any.whl
Collecting pyasn1<0.5.0,>=0.4.6
  Using cached https://files.pythonhosted.org/packages/62/1e/a94a8d635fa3ce4cfc7f506003548d0a2447ae76fd5ca53932970fe3053f/pyasn1-0.4.8-py2.py3-none-any.whl
Collecting oauthlib>=3.0.0
  Using cached https://files.pythonhosted.org/packages/05/57/ce2e7a8fa7c0afb54a0581b14a65b56e62b5759dbc98e80627142b8a3704/oauthlib-3.1.0-py2.py3-none-any.whl
ERROR: phoebe 2.1.15 requires scipy>=0.17, which is not installed.
Installing collected packages: numpy, opt-einsum, grpcio, absl-py, wrapt, h5py, keras-applications, keras-preprocessing, google-pasta, pyasn1, pyasn1-modules, rsa, cachetools, google-auth, markdown, protobuf, werkzeug, urllib3, idna, chardet, requests, oauthlib, requests-oauthlib, google-auth-oauthlib, tensorboard, gast, termcolor, astor, tensorflow-estimator, tensorflow
Successfully installed absl-py-0.9.0 astor-0.8.1 cachetools-4.0.0 chardet-3.0.4 gast-0.2.2 google-auth-1.11.2 google-auth-oauthlib-0.4.1 google-pasta-0.1.8 grpcio-1.27.2 h5py-2.10.0 idna-2.8 keras-applications-1.0.8 keras-preprocessing-1.1.0 markdown-3.2.1 numpy-1.18.1 oauthlib-3.1.0 opt-einsum-3.1.0 protobuf-3.11.3 pyasn1-0.4.8 pyasn1-modules-0.2.8 requests-2.22.0 requests-oauthlib-1.3.0 rsa-4.0 tensorboard-2.0.2 tensorflow-2.0.0 tensorflow-estimator-2.0.1 termcolor-1.1.0 urllib3-1.25.8 werkzeug-1.0.0 wrapt-1.12.0

Installation complete.


Restarting R session...

**Comment by gtatters** [comment]
then I repeat the original request from earlier:

```
tensorflow::tf_config()
```

And I get the same tensorflow problem before:

TensorFlow v2.0.0 (~/.virtualenvs/r-reticulate/lib/python3.7/site-packages/tensorflow)
Python v3.7 (~/.virtualenvs/r-reticulate/bin/python)

Isn't it supposed to use the conda I just installed?    Is there a command I am supposed to use to force it to find the miniconda I just installed?

**Comment by skeydan** [comment]
Can you restart R, then do

```
reticulate::use_condaenv("yourenvname", required = TRUE)
```

This should error if it cannot switch to that env.

**Comment by gtatters** [comment]
I can, but what is "yourenvname"?  This implies I have to set an environment and I haven't seen that.

**Comment by gtatters** [comment]
```
reticulate::use_condaenv(conda="/Users/USERNAME/opt/miniconda3/bin/conda")
```

produces no error or output.  I have changed my actual username to USERNAME to avoid my name being copied and pasted.

**Comment by gtatters** [comment]
whereas 
```
reticulate::use_condaenv(conda="yourenvname", required=TRUE)
```
Yields the following error, not surprisingly:

Error in reticulate::use_condaenv("yourenvname", required = TRUE) : 
  Unable to locate conda environment 'yourenvname'.

**Comment by skeydan** [comment]
oh, sorry, I meant whatever you called that env, but as you didn't name it explicitly it should be 

`r-reticulate`

**Comment by gtatters** [comment]
yeah...Keras gave no instructions on this.  So am I supposed to create an environment?  How do I do that?

**Comment by gtatters** [comment]
anyhow, 
```
reticulate::use_condaenv("r-reticulate", required = TRUE)
```
provides no error or output.  seems to be happy.

**Comment by gtatters** [comment]
So, now, after install_tensorflow(method="virtualenv") is run and then 
tensorflow::tf_config() is run, I still get:

TensorFlow v2.0.0 (~/.virtualenvs/r-reticulate/lib/python3.7/site-packages/tensorflow)
Python v3.7 (~/.virtualenvs/r-reticulate/bin/python)

**Comment by skeydan** [comment]
Sorry, I don't follow... Why did you run `install_tensorflow(method="virtualenv") ` now?

Going back to the conda env, you say that

```
reticulate::use_condaenv("r-reticulate", required = TRUE)
```

throws no error... so what does `tf_config()` say when run after that command (and in a fresh R session)?

**Comment by gtatters** [comment]
restarting everything Rstudio and R and then running 

```
reticulate::use_condaenv("r-reticulate", required = TRUE)
```
then running:

```
tensorflow::tf_config()
```

yields:

> Error in sitecustomize; set PYTHONVERBOSE for traceback:
> KeyError: 'PYTHONPATH'
> Installation of TensorFlow not found.
> 
> Python environments searched for 'tensorflow' package:
>  /Users/USERNAME/opt/miniconda3/envs/r-reticulate/bin/python3.6
> 
> You can install TensorFlow using the install_tensorflow() function.

**Comment by skeydan** [comment]
Hm, this looks like a possible problem with Anaconda on Mac, do any of these help?

http://agapow.net/programming/python/verbose/

https://stackoverflow.com/questions/32321973/after-installing-anaconda-i-get-constant-keyerror-pythonpath-messages

https://groups.google.com/a/continuum.io/forum/#!msg/anaconda/IX4GxgCIkqQ/aRLuKZHQDgAJ

**Comment by gtatters** [comment]
Thank you.  I will try that.  Having python2 installed in mac by default, then adding python3 via homebrew, then installing miniconda (per keras suggestions), there seems to be some confusion on my OS.  I will let you know the outcome once I have success with wrestling the python install.  I thought it would all work "virtually" without needing python, thus why I posted to the issues here.

**Comment by gtatters** [comment]
So, the error was the multiple versions of python on my OS and confusion with .bash_profile most likely.  Removing, unlinking, and uninstalling and reinstalling python in homebrew was required.

I mostly followed this site: 

https://www.pyimagesearch.com/2019/01/30/macos-mojave-install-tensorflow-and-keras-for-deep-learning/

except that I installed python 3.7 since it looks compatible with tensorflow now.

Sorry for all the questions.  At least I've got past loading the data and am training the model now.  Looking forward to seeing this work!
